File: webapp/src/api/http/validation.ts
/**
 * The error with a particular field
 */
export class FieldError {
  /** The field that the error occurred with */
  readonly field: string;
  /** The actual error type that occurred */
  readonly type: string;

  /**
   * Constructor
   * @param field The name of the field
   * @param type The error type
   */
  constructor(field: string, type: string) {
    this.field = field;
    this.type = type;
  }
}

/**
 * A validation error occurred on a request
 */
export class ValidationError {
  /** The fields that were in error */
  readonly fields: FieldError[];

  /**
   * Constructor
   * @param fields The field that were in error
   */
  constructor(fields: FieldError[]) {
    this.fields = fields;
  }

  /**
   * Get all of the errors with a particular field
   * @param field The name of the field
   */
  getFieldErrors(field: string) {
    return this.fields.filter((f) => f.field === field);
  }
}

File: service/crates/http/src/problem/validation.rs
use super::{Problem, SimpleProblemType};
use actix_http::http::StatusCode;
use serde::Serialize;
use std::fmt::{Debug, Display};

pub const VALIDATION_ERROR: SimpleProblemType = SimpleProblemType {
    problem_type: "tag:mire/2020:problems/validation_error",
    problem_title: "A validation error occurred",
    status_code: StatusCode::UNPROCESSABLE_ENTITY,
};

/// Trait to represent the type of field validation error.
pub trait FieldValidationType: Debug + Display {
    /// A URI Reference that identifies the field validation type.
    fn problem_type(&self) -> &'static str;
}

/// Details of a validation error on a field
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FieldValidation {
    /// The name of the field in error
    field_name: String,
    /// The actual error
    r#type: String,
    /// The human readable description of the error
    title: String,
}

/// Builder for constructing a validation error problem
#[derive(Debug, Default)]
pub struct ValidationErrorBuilder {
    /// The fields that are in error
    fields: Vec<FieldValidation>,
}

impl ValidationErrorBuilder {
    /// Record a new problem with a field
    ///
    /// # Parameters
    /// - `field_name` - The name of the field in error
    /// - `problem` - The problem with the field
    pub fn with_field<S, F>(&mut self, field_name: S, problem: &F) -> &mut Self
    where
        S: Into<String>,
        F: FieldValidationType,
    {
        let r#type = problem.problem_type().to_owned();
        let title = problem.to_string();

        self.fields.push(FieldValidation {
            field_name: field_name.into(),
            r#type,
            title,
        });

        self
    }

    /// Build a problem from the builder
    #[must_use]
    pub fn build(self) -> Problem {
        Problem::new(VALIDATION_ERROR).with_extra("fields", self.fields)
    }
}
